lib/active_shipping/carriers/ups.rb

```diff
@@ -961,6 +977,95 @@ def parse_void_response(response, options={})
       end
     end
 
+    def build_address_validation_request(location, options = {})
+      xml_builder = Nokogiri::XML::Builder.new do |xml|
+        xml.AddressValidationRequest do
+          xml.Request do
+            xml.RequestAction('XAV')
+            xml.RequestOption('3')
+
+            if options[:customer_context]
+              xml.TransactionReference do
+                xml.CustomerContext(options[:customer_context])
+                xml.XpciVersion("1.0")
+              end
+            end
+          end
+
+          xml.AddressKeyFormat do
+            xml.AddressLine(location.address1)
+            if location.address2.present?
+              xml.AddressLine(location.address2)
+            end
+            xml.PoliticalDivision2(location.city)
+            xml.PoliticalDivision1(location.state)
+            xml.PostcodePrimaryLow(location.postal_code)
+            xml.CountryCode(location.country_code)
+          end
+        end
+      end
+
+      xml_builder.to_xml
+    end
+
+    def parse_address_validation_response(address, response, options={})
+      xml     = build_document(response, 'AddressValidationResponse')
+      success = response_success?(xml)
+      message = response_message(xml)
+
+      validity = nil
+      classification_code = nil
+      classification_description = nil
+      addresses = []
+
+      if success
+        classification_code = xml.at('AddressClassification/Code').text
+        
+        if classification_code == "1"
+          classification = :commercial
+        elsif classification_code == "2"
+          classification = :residential
+        else
+          classification = :unknown
+        end
+
+        if xml.at("ValidAddressIndicator")
+          validity = :valid
+        elsif xml.at("AmbiguousAddressIndicator")
+          validity = :ambiguous
+        elsif xml.at("NoCandidatesIndicator")
+          validity = :invalid
+        else
+          validity = :unknown
+        end
+
+        xml.css('AddressKeyFormat').each do |address_key_node|
+          addresses << location_from_address_key_format_node(address_key_node)
+        end
+      end
+
+      response = AddressValidationResponse.new(success, message, Hash.from_xml(response).values.first, :validity => validity, :classification => classification, :candidate_addresses => addresses, :xml => response, :request => last_request)
```

Comment:
Could you build out `params = Hash.from_xml(response).values.first` outside of the `AddressValidationResponse` construction? It's an added variable initialization but I feel like a little goes a long way in understanding exactly what you're trying to pass here, which is `params`.